### PR TITLE
JS: Handle sub-dep version resolution errors

### PR DESCRIPTION
**Pull request by feelepxyz**
Raise `Dependabot::InconsistentRegistryResponse` if the version can't be
found for both top level and sub-dependencies, including
sub-dependencies that are not yet known to Dependabot.

Error is handled in the updater when creating PRs.

### CODE REVIEW

**Comment by feelepxyz** [comment]
@greysteil just pushed the changes with a spec - included a somewhat random cleanup which I thought I needed in the error handling.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb [hunk 1]
```diff
@@ -193,27 +191,64 @@ def handle_npm_updater_error(error, lockfile)
...
+            # the case where a new version (e.g. @angular-devkit/build-angular)
+            # relies on a added dependency which hasn't been published yet under
+            # the same scope (e.g. @angular-devkit/build-optimizer)
+            #
+            # This seems to happen when big monorepo projects publish all of
+            # their packages sequentially, which might take enough time for
+            # Dependabot to hear about a new version before all of its
+            # dependencies have been published.
+            names = dependencies.map(&:name).map do |name|
+              name.split("/").first
```
Comment (greysteil) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb]: `names = dependencies.map { |dep| dep.name.split("/").first }` to go from three lines to one.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb [hunk 2]
```diff
@@ -193,27 +191,62 @@ def handle_npm_updater_error(error, lockfile)
...
+            # the case where a new version (e.g. @angular-devkit/build-angular)
+            # relies on a added dependency which hasn't been published yet under
+            # the same scope (e.g. @angular-devkit/build-optimizer)
+            #
+            # This seems to happen when big monorepo projects publish all of
+            # their packages sequentially, which might take enough time for
+            # Dependabot to hear about a new version before all of its
+            # dependencies have been published.
+            names = dependencies.map { |dep| dep.name.split("/").first }
+            if names.any? { |name| error.message.include?(name) }
```
Comment (greysteil) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb]: `error.message.match?(/Regexp.quote(name)[/@]/)` so as the name has to either match a dependency's scope or the end of its name.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb [hunk 3]
```diff
@@ -193,27 +191,62 @@ def handle_npm_updater_error(error, lockfile)
...
              error.message.include?("Invalid tag name")
-            # This happens if a new version has been published that relies on
-            # subdependencies that have not yet been published.
-            raise if resolvable_before_update?(lockfile)
+            unless resolvable_before_update?(lockfile)
+              raise_resolvability_error(error, lockfile)
+            end
 
-            raise_resolvability_error(error, lockfile)
+            # TODO: Move this logic to the version resolver and check if a new
```
[the remark this answers is not in the document]
Comment (feelepxyz) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb]: Ah do we not always want to raise `raise_resolvability_error` if it's not resolvable before update? Which would mean there's something wrong with the lockfile and re-running the update will probably never work?

**Pull request merged**